### ISSUE TITLE
Device-side matrix assembly: device_matrix_data::sum_duplicates

### ISSUE THREAD

**Issue by upsj** [ginkgo-project/ginkgo#946]
For a purely device-side matrix assembly, we have no nice interface (also since they would be executor-dependent). I was thinking that we could use device_matrix_data for this, by adding a function that sorts the input in row-major, sums up the resulting adjacent duplicates into a single entry and stores it in the device matrix data again, structurally similar to what remove_zeros does.

Does something like that sound useful? Especially asking people with application experience here, @nbeams @pratikvn @Slaedr @greole @fritzgoebel @MarcelKoch 

Also there is the common question of naming. Ideas that came to mind:

* sum_duplicates: mentions the reduction and its operation, but does not explicitly mention resulting uniqueness
* add_duplicates: like sum_duplicates, but add could also mean *insert new entries*
* remove_duplicates: only focuses on uniqueness
* combine_duplicates: mentions uniqueness and (implied) reduction, but not exact reduction operation
* reduce_duplicates: like combine_duplicates, just more explicit on the reduction part

### Discussed in https://github.com/ginkgo-project/ginkgo/discussions/945

<div type='discussions-op-text'>

<sup>Originally posted by **edocento** December 30, 2021</sup>
I cannot find a way to fill a larger matrix (`Coo` or `Csr` type) with smaller blocks at specific coordinates in such a way that values added in previous steps are summed up to new values at the same (row,col) coordinate. In other words, I would like to reproduce with your library a function like `MatSetValuesLocal(...)` in Petsc (with flag `ADD_VALUES`) . 
(Here the docs of the function I am referring to https://petsc.org/main/docs/manualpages/Mat/MatSetValuesLocal.html )</div>

**Comment by MarcelKoch** [comment]
IMO having that capability would be a good thing, and I wouldn't expect it to be too difficult to implement. However, using that as a substitute for `matix_assembly_data` might be a bit of a stretch. One issue I see is that you would need to know the total number of matrix entries, _including_ duplicated, beforehand. It should be easily possible to compute an overestimate, but getting the exact number should be quite hard.

**Comment by upsj** [comment]
Yes, I fully agree there! The example I had in mind was again FEM assembly, where you allocate num_elements * nodes_per_element storage and assemble all elements independently.

**Comment by Slaedr** [comment]
I think that's indeed something we could support. While it is reasonable to say that we mainly target large-scale scientific applications which would typically want to build their matrices in a way that's agnostic of the linear algebra library (thus we have array view constructors for our matrix types), this would be needed if we want to support other kinds of users that don't already have an established application codebase and are open to using Ginkgo exclusively.

At least for an initial but usable implementation, we can expect the user to provide an (over) estimate of the total number of nonzeros + duplications.  However, I'm not sure how we can provide a device-side interface for users to set the nonzero triplets. Is the idea to let insertion be the user's problem, and we just take a Ginkgo `Array<matrix_data_entry>` to construct the `device_matrix_data`?

**Comment by upsj** [comment]
yes, the users would need to take care of indexing themselves. If you want `insert` functionality without indexing, you'd need an individual (over)estimate for each threadblock(GPU)/core(CPU) or a two-pass approach (counting and then filling), and I don't want to overcomplicate things here.
